### PR TITLE
fix(DeASDA): Add support for A3 series of drives

### PR DESCRIPTION
**Pull request by GuiHue**
fix(DeASDA): Add support for A3 series of drives

This adds support for Delta ASD-A3 series of drives.

Tested for function (Slave going OP + drive moving) on actual hardware
Deliberatly did not rename DeASDA to DeASDA2 as to not break existing configs.

 Issue: #199

### CODE REVIEW

**Review by scottlaird — APPROVED** [review]

**Comment by GuiHue** [comment]
@scottlaird Hi, I have reworked a lot since last PR:
-Added default pprev for A2 and A3
-made A2 default to assure compatibility with legacy sittner/linuxcnc-ethercat configs
-Added modParam opmode with values CSV and CSP
-Tested against a dual setup of 1 A3 and 1 A2 series drive on my desk.

Noch this PR addresses issues #201  and #199

**Comment by GuiHue** [comment]
@scottlaird This one seems to be stuck. Anything else I can do? Development on the driver has finished from my side. Basic Docs are provided.
I am not too familiar with pull requests. It seems that this one is a bit screwed up since I made later changes in the same branch. This branch now contains the complete version.

**Comment by scottlaird** [comment]
> @scottlaird This one seems to be stuck. Anything else I can do? Development on the driver has finished from my side. Basic Docs are provided. I am not too familiar with pull requests. It seems that this one is a bit screwed up since I made later changes in the same branch. This branch now contains the complete version.

Ah, sorry--for some reason I didn't think it was ready.  I'll take a look.

**Review by scottlaird — CHANGES_REQUESTED** [review]
Generally fine, I have a handful of style nits, feel free to take them or leave them.

The only thing that I'd really like to see changed is `creating-ethercat-configs.md`.  As it stands, it's not really useful.  Either drop it from this PR or add something there?

File: examples/creating-ethercat-configs.md
Comment (GuiHue) [reply examples/creating-ethercat-configs.md:9]: Dropped - was never meant to be part of the commit... my mistake.

File: src/devices/lcec_deasda.c
Comment (GuiHue) [reply src/devices/lcec_deasda.c:129]: While I cannot merge the similar pins into slave_param (see pins vs. params), I managed to split slave_pins_csv /csp into slaves_pins for the common ones and .._csp/csv respectively for the differentiating pins. Hopefully this makes stuff cleaner.

File: src/devices/lcec_deasda.c
Comment (GuiHue) [reply src/devices/lcec_deasda.c:287]: Currently, only these two modes are supported and they seem to be the most relevant. However, I'd like to keep the option open. Repeated use is required as the distinction in the modes is required at multiple points. However, I have adopted the defines to make stuff more readable

Comment (GuiHue) [reply src/devices/lcec_deasda.c:287]: Yes, both A2 and A3 can go into both CSV and CSP. The only actual difference for the time being is the default encoder setting. hence no changes are required here.

File: src/devices/lcec_deasda.c
Comment (GuiHue) [reply src/devices/lcec_deasda.c:300]: Thanks. You're right. Done.

File: src/devices/lcec_deasda.c
Comment (GuiHue) [reply src/devices/lcec_deasda.c:286]: Installed, tried and I think it didn't mess up too badly.

**Comment by GuiHue** [comment]
@scottlaird Thank you for the detailed feedback. Greatly appreciated. i hope to have addressed those topics. The offending file has been removed, improvements were made and the code tested on the bench. Hope this is good to go now;).

**Review by scottlaird — APPROVED** [review]

**Comment by scottlaird** [comment]
Looks good, thanks.  Go ahead and hit merge whenever you're ready :-).

**Comment by scottlaird** [comment]
FYI, I just merged #244 with a bulk edit to all SDO handling code, and didn't realize that it touched the deasda driver until after it went it.  You'll probably need to tweak the SDO setting code very slightly; see #243 for the reasoning.

**Comment by GuiHue** [comment]
@scottlaird 
Dang, I should have chosen a different profession/major other than mechanical engineering. I need a good online course on git to get a better understanding. I think I have done all the right steps but I am not 100% sure.

**Comment by scottlaird** [comment]
> @scottlaird Dang, I should have chosen a different profession/major other than mechanical engineering. I need a good online course on git to get a better understanding. I think I have done all the right steps but I am not 100% sure.

Yeah, git can be fun...  It's not too bad if you stick the the simple stuff, but as soon as you leave the easy path it gets *weird*.  I've been using [Graphite](http://graphite.dev) lately; it's a layer on top of Git that focuses on managing stacks of related changes and *mostly* makes things better.  Except when it decides that you need to drop down into git because you hit something that they haven't implemented yet.

**Comment by scottlaird** [comment]
> > @scottlaird Dang, I should have chosen a different profession/major other than mechanical engineering. I need a good online course on git to get a better understanding. I think I have done all the right steps but I am not 100% sure.
> 
> Yeah, git can be fun... It's not too bad if you stick the the simple stuff, but as soon as you leave the easy path it gets _weird_. I've been using [Graphite](http://graphite.dev) lately; it's a layer on top of Git that focuses on managing stacks of related changes and _mostly_ makes things better. Except when it decides that you need to drop down into git because you hit something that they haven't implemented yet.

FWIW, it's not clear to me why this isn't building, but I can reproduce it.  I'll probably have a suggestion in a few minutes.

File: src/devices/lcec_deasda.c
Comment (scottlaird) [inline src/devices/lcec_deasda.c:200]: I *think* this `}` needs to stay in the file; that's what's causing the build failure.

File: src/devices/lcec_deasda.c
Comment (scottlaird) [inline src/devices/lcec_deasda.c:265]: add a `return -1;` here (and a couple lines below) so that LinuxCNC will fail to start if SDO setting fails?

**Review by scottlaird — APPROVED** [review]

**Comment by GuiHue** [comment]
Damn - i may have missed something during merging. Looked fine initially. I'll tackle it tonight. Sorry about that.

File: src/devices/lcec_deasda.c
Comment (GuiHue) [reply src/devices/lcec_deasda.c:266]: added for multiple locations.

File: src/devices/lcec_deasda.c
Comment (GuiHue) [reply src/devices/lcec_deasda.c:200]: Your thinking was right;). Also, during the merge another bad bad occurred that disabled CSP mode. Fixed and tested.

**Comment by GuiHue** [comment]
@scottlaird Fixes applied - hopefully one last time;)

**Comment by scottlaird** [comment]
Looks good, thanks.  I'll go ahead and merge.

**Comment by scottlaird** [comment]
Okay, I'm mystified here.  Can you try hitting the merge button from your end?  Maybe it's trying to modify your branch and I don't have permissions there?  I don't see any obvious conflicts.

**Comment by GuiHue** [comment]
trying... I get all greens on this side. Can't hit merge though as it says:
"This branch has no conflicts with the base branch
Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests."
Compile locally works fine as well. My branch in my fork says
"This branch is not behind the upstream linuxcnc-ethercat/linuxcnc-ethercat:master.
No new commits to fetch. Enjoy your day!"

**Comment by scottlaird** [comment]
Hmm. Okay, that's on me, then. I'll figure it out.

**Comment by scottlaird** [comment]
I *think* it was a combination of the merge queue being enabled and 'rebase and merge' being the default.  Should be done now.

**Comment by GuiHue** [comment]
You rock! Thanks mate!

**Pull request merged**
